Add setting to localize numeric dates independent from language (#15770)

christophwen opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change

<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue or discussion
  in the additional information section.
-->

Issue #15770 

Problem: Language set to English used by numeric date formatting -> result 5/9/2023 (M/D/YY used with en-US).
Wanted: Language set to English, but date should be displayed as DD/MM/YY (like en-GB)
Solution: Numeric date format can be changed with a profile setting independent of the set language

Setting in user profile:

![User profile](https://github.com/home-assistant/frontend/assets/43651938/4109f3d1-b8d1-4827-9317-9a0c52504d69)

Effect: 

![Adjust a statistic](https://github.com/home-assistant/frontend/assets/43651938/af50feb5-ea0a-468e-80fa-111447c7ea53)


## Type of change

<!--
  What type of change does your PR introduce to the Home Assistant frontend?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR.
-->

```yaml

```

## Additional information

<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #15770 
- This PR is related to issue or discussion: #15770 
- Link to documentation pull request:

## Checklist

<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

<!--
  Thank you for contributing <3
-->

[docs-repository]: https://github.com/home-assistant/home-assistant.io

bramkragten commented:
This should just be `Date format`, not `Numeric date format`

bramkragten commented on src/data/translation.ts:
```diff
@@ -17,6 +17,12 @@ export enum TimeFormat {
 
+export enum NumericDateFormat {
+  language = "language",
+  MDY = "MDY",
+  DMY = "DMY",
```
Why so little options? There are a lot more possibilities? Like I'm in the Netherlands, if I want the UI in English, but the dates in my normal format I would use `D-M-YY` not `DD/MM/YY`.

I would at least expect a system option here, like we have for the others

bramkragten commented:
Btw, the solution for your problem is already there, we have 2 english languages, `English` and `English (GB)` that does exactly what you want.

bramkragten commented:
Change the language in your HA profile to `English (GB)`

christophwen commented:
> This should just be `Date format`, not `Numeric date format`

At first i used `Date format` but because it only affects the numeric date (formatter) I decided to `Numeric date format`.
All other date formats are not affected by this setting. Should I change it anyway?

> Why so little options? There are a lot more possibilities? Like I'm in the Netherlands, if I want the UI in English, but the dates in my normal format I would use D-M-YY not DD/MM/YY.
> 
> I would at least expect a system option here, like we have for the others

Thats because I couldn't figure out if and how I can use `Intl.DateTimeFormat` and make custom formatting. 
I can custom format the date without using `Intl.DateTimeFormat` as a solution and add more options (+ system).

> Btw, the solution for your problem is already there, we have 2 english languages, English and English (GB) that does exactly what you want.

Yes, you are right. But I wanted to make it more flexible and independent (if a setting is selected).
Scenario: You want American English but a not American date format. (Like me, who lives in Germany and prefers American English. Don't ask why, I don't know myself. ^^)

PS.: I will fixed the `lint:prettier` warnings. It is my first PR and I forgot about that, sorry.

ildar170975 commented:
> we have 2 english languages, `English` and `English (GB)` that does exactly what you want.

There is another case - Russian language + desired format DD/MM/YY.
So, making date format dependent on language is a wrong way.

christophwen commented:
> > we have 2 english languages, `English` and `English (GB)` that does exactly what you want.
> 
> There is another case - Russian language + desired format DD/MM/YY. So, making date format dependent on language is a wrong way.

I agree, but I think using the language as a default is okay.

ildar170975 commented:
> using the language as a default is okay

Agree! And user should have a choice of selecting another date format. Hope this is the idea of this PR.

christophwen commented:
>  Hope this is the idea of this PR.

Yes it is. Do you think it is not clearly described above?

bramkragten commented:
> At first i used Date format but because it only affects the numeric date (formatter) I decided to Numeric date format.
All other date formats are not affected by this setting. Should I change it anyway?

I don't think we should ever have multiple options for date formatting, so yes I would change it

> Thats because I couldn't figure out if and how I can use Intl.DateTimeFormat and make custom formatting.
I can custom format the date without using Intl.DateTimeFormat as a solution and add more options (+ system).

We can just link options to different locales, like we also do for number formatting:
https://github.com/home-assistant/frontend/blob/dev/src/common/number/format_number.ts#L24-L31

bramkragten commented:
btw, don't forget to also implement this for `formatDateTimeNumeric`

bramkragten commented:
> Thats because I couldn't figure out if and how I can use Intl.DateTimeFormat and make custom formatting.
I can custom format the date without using Intl.DateTimeFormat as a solution and add more options (+ system).

A way to do that would be to use `formatToParts`:
```ts
const parts = new Intl.DateTimeFormat("en-US", {
      day: "numeric",
      month: "short",
    }).format(date); // May 10
```

```ts
const parts = new Intl.DateTimeFormat("en-US", {
      day: "numeric",
      month: "short",
    }).formatToParts(date);
return `${parts[2].value} ${parts[0].value}`; // 10 May
```

christophwen commented:
> I don't think we should ever have multiple options for date formatting, so yes I would change it

Changed ✓

> I would at least expect a system option here, like we have for the others

Added ✓

> Maybe we should only give 3 options, YMD, DMY and MDY.

Currently possible options/combinations (which make sense in my opinion):

- "language": "Auto (use language setting)"
- "system": "Use system locale"
- "MDYY": "M/D/YY"
- "MMDDYY": "MM/DD/YY"
- "DMYY": "D/M/YY"
- "DDMMYY": "DD/MM/YY"
- "YYMD": "YY/M/D"
- "YYMMDD": "YY/MM/DD"

> btw, don't forget to also implement this for formatDateTimeNumeric

Fixed ✓


> A way to do that would be to use formatToParts

Thank you, that was very useful!

bramkragten commented:
> Currently possible options/combinations (which make sense in my opinion):
>
> "language": "Auto (use language setting)"
"system": "Use system locale"
"MDYY": "M/D/YY"
"MMDDYY": "MM/DD/YY"
"DMYY": "D/M/YY"
"DDMMYY": "DD/MM/YY"
"YYMD": "YY/M/D"
"YYMMDD": "YY/MM/DD"

Ooff... now I think a lot of people are gonna feel left out, I think al those variant also exist with a `.` or `-` instead of the `/`. That's why I proposed to only offer 3, and use that as the order. Don't change the formatting itself but only the order of the day, month and year?

christophwen commented:
> Ooff... now I think a lot of people are gonna feel left out, I think al those variant also exist with a `.` or `-` instead of the `/`. That's why I proposed to only offer 3, and use that as the order. Don't change the formatting itself but only the order of the day, month and year?

You are right, I will change that tomorrow. Sorry for all the revisions. ^^

bramkragten commented on src/common/datetime/format_date.ts:
```diff
@@ -32,15 +32,55 @@ const formatDateMem = memoizeOne(
+      day: "numeric",
+    }).formatToParts(dateObj);
+
+    const literal = parts
+      .filter((value) => value.type === "literal")
+      .map(({ value }) => value)[0];
```
```suggestion
    const literal = parts
      .find((value) => value.type === "literal").value;
```

bramkragten commented on src/common/datetime/format_date.ts:
```diff
@@ -32,15 +32,55 @@ const formatDateMem = memoizeOne(
+    }
+
+    let result = new Intl.DateTimeFormat(localeString, {
       year: "numeric",
       month: "numeric",
       day: "numeric",
-    })
+    }).format(dateObj);
```
If system or use language we can just return here, no need to do all the stuff below

bramkragten commented:
Can we add a page to the gallery to prints all options for all our supported languages in a table?

christophwen commented:
> Can we add a page to the gallery that prints all options for all our supported languages in a table?

Done.

I also fixed some formates for a few languages.

christophwen committed:
Add setting to localize numeric dates independent from language
